src/poprox_recommender/components/joiners/fill.py

```diff
@@ -0,0 +1,15 @@
+from poprox_concepts import ArticleSet
+
+
+class Fill:
+    def __init__(self, num_slots):
+        self.num_slots = num_slots
+
+    def __call__(self, candidates1: ArticleSet, candidates2: ArticleSet) -> ArticleSet:
+        # Fill as many slots as you can from candidates1
+        articles = candidates1.articles[: self.num_slots]
+
+        # Fill out remaining slots from candidates2
+        articles += candidates2.articles[: self.num_slots - len(articles)]
```

Comment:
what happens when the index is negative here? (I do not remember that nuance of Python indexing.)

Reply:
For consideration — would it be easier with itertools? `list(islice(chain(candidates1.articles, candidates2.articles), self.num_slots))` — then all the edge cases around the lengths of the two article lists are handled by the library.

I don't have a strong opinion on relative readability, but making the code “obviously correct” in the face of edge cases (the key one being when `candidates1` has more than `num_slots` articles) may add complexity.

Reply:
I'm generally good with improving the edge case handling of this code but it's not obvious to me what the itertools snippet is actually doing without the itertools docs in front of me 😅

Reply:
At the minor cost of more list allocations, we could do list combination:

```python
articles = candidates1.articles + candidates2.articles
articles = articles[:self.num_slots]
```